Optionally soft-link WebContentAnalysis.framework and WebFilterEvaluator

whsieh opened the pull request:
#### b5a5e0bd3bb7c037b0081b67c21a772226a70187
<pre>
Optionally soft-link WebContentAnalysis.framework and WebFilterEvaluator
<a href="https://bugs.webkit.org/show_bug.cgi?id=251947">https://bugs.webkit.org/show_bug.cgi?id=251947</a>

Reviewed by Aditya Keerthi.

Don&apos;t `RELEASE_ASSERT` in the case where `WebContentAnalysis.framework` is missing.

* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:

Canonical link: <a href="https://commits.webkit.org/260045@main">https://commits.webkit.org/260045@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/350692957b76aa8e809ec279f74bb434ae578265

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106829 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15858 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39617 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/116012 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110732 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/17344 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/7069 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/99016 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112597 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/13138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/96138 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40738 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95022 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/27790 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9025 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/29145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9596 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/6156 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/15193 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48690 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6934 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/11123 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/cocoa/ContentFilterUnblockHandlerCocoa.mm:
```diff
@@ -118,7 +119,8 @@
 #if HAVE(PARENTAL_CONTROLS_WITH_UNBLOCK_HANDLER)
-    unblockHandler.m_webFilterEvaluator = [coder decodeObjectOfClass:getWebFilterEvaluatorClass() forKey:webFilterEvaluatorKey];
+    if (WebContentFilterLibrary())
+        unblockHandler.m_webFilterEvaluator = [coder decodeObjectOfClass:webFilterEvaluatorClass forKey:webFilterEvaluatorKey];
```
Shouldn't this still be `getWebFilterEvaluatorClass()`? I don't see where `webFilterEvaluatorClass` is defined.

whsieh replied:
> Shouldn't this still be `getWebFilterEvaluatorClass()`? I don't see where `webFilterEvaluatorClass` is defined.

🤦 yes (I originally had `auto webFilterEvaluatorClass  = getWebFilterEvaluatorClass()`, changed it to `WebContentFilterLibrary()` and forgot to change the `webFilterEvaluatorClass`)

whsieh replied:
> > Shouldn't this still be `getWebFilterEvaluatorClass()`? I don't see where `webFilterEvaluatorClass` is defined.
> 
> 🤦 yes (I originally had `auto webFilterEvaluatorClass = getWebFilterEvaluatorClass()`, changed it to `WebContentFilterLibrary()` and forgot to change the `webFilterEvaluatorClass`)

I undid all of my changes to this file since it's all `PLATFORM(IOS_FAMILY)` only.

pxlcoder commented on Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:
```diff
@@ -36,8 +36,8 @@
 
-SOFT_LINK_PRIVATE_FRAMEWORK(WebContentAnalysis);
-SOFT_LINK_CLASS(WebContentAnalysis, WebFilterEvaluator);
+SOFT_LINK_PRIVATE_FRAMEWORK_OPTIONAL(WebContentAnalysis);
```
Are there no `WebContentFilterLibrary()` checks here, because `WebContentAnalysis` is only used on iOS? Should this soft linking also be restricted in that case?

pxlcoder replied:
Or does the logic simply not run in places where the framework is missing?

pxlcoder replied:
Never mind, I suppose `[getWebFilterEvaluatorClass() isManagedSession]` will just evaluate to false, which is fine.

whsieh replied:
It should be fine here, since `SOFT_LINK_CLASS_OPTIONAL` will make it so that `getWebFilterEvaluatorClass()` returns `nil` (which the rest of this code handles gracefully).

However, this comment made me realize that we don't actually need to touch `ContentFilterUnblockHandler` at all, since the base system isn't a thing on iOS. I'll remove my changes there to minimize this diff.

whsieh committed:
Optionally soft-link WebContentAnalysis.framework and WebFilterEvaluator
https://bugs.webkit.org/show_bug.cgi?id=251947

Reviewed by Aditya Keerthi.

Don't `RELEASE_ASSERT` in the case where `WebContentAnalysis.framework` is missing.

* Source/WebCore/platform/cocoa/ParentalControlsContentFilter.mm:

Canonical link: https://commits.webkit.org/260045@main
